Added a "How to convert your Depends dependency on data.table to Imports" section to datatable-importing vignette. (#6161)
Rdatatable/data.table: vignettes/datatable-importing.Rmd

Some of internally used C routines are now exported on C level thus can be used in R packages directly from their C code. See [`?cdt`](https://rdatatable.gitlab.io/data.table/reference/cdt.html) for details and [Writing R Extensions](https://cran.r-project.org/doc/manuals/r-release/R-exts.html) _Linking to native routines in other packages_ section for usage.

## Importing from non-r Applications {non-r-api}

Some tiny parts of `data.table` C code were isolated from the R C API and can now be used from non-R applications by linking to .so / .dll files. More concrete details about this will be provided later; for now you can study the C code that was isolated from the R C API in [src/fread.c](https://github.com/Rdatatable/data.table/blob/master/src/fread.c) and [src/fwrite.c](https://github.com/Rdatatable/data.table/blob/master/src/fwrite.c).

## How to convert your Depends dependency on data.table to Imports

To convert a `Depends` dependency on `data.table` to an `Imports` dependency in your package, follow these steps:

### Step 0.  Ensure your package is passing R CMD check initially

### Step 1. Update the DESCRIPTION file to put data.table in Imports, not Depends

**Before:**

> ```r

Reviewer: ```suggestion
```
```

> ```r

Reviewer: remove the `r` (this is not R syntax)

you can try ```` ```dcf ````, not sure that's recognized language for highlighting by any service

Depends:
    R (>= 3.5.0),
    data.table
Imports:
```

**After:**
```r
Depends:
    R (>= 3.5.0)
Imports:
    data.table
```

### Step 2.1: Run `R CMD check`

Run `R CMD check` to identify any missing imports or symbols. This step helps:

- Automatically detect any functions or symbols from `data.table` that are not explicitly imported.
- Flag missing special symbols like `.N`, `.SD`, and `:=`.
- Provide immediate feedback on what needs to be added to the NAMESPACE file.

> 

Reviewer: We should add a caveat here: not all such usages are caught by `R CMD check`. In particular `R CMD check` skips some symbols/functions in formulas, and will completely miss parsed expressions like `parse(text = "data.table(a = 1)")`. Packages will need good test coverage to detect these edge cases.

### Step 2.2: Modify the NAMESPACE File

> Based on the `R CMD check` results, ensure all used functions, special symbols, S3 generics, and S4 classes from `data.table` are imported.

Reviewer: ```suggestion
Based on the `R CMD check` results, ensure all used functions, special symbols, S3 generics, and S4 classes from `data.table` are imported.

That means adding `importFrom(data.table, ...)` and/or `importClassesFrom(data.table, ...)` directives as appropriate. See 'Writing R Extensions' for full details on how to do so properly.
```

Reviewer: I think the below `####` sections are overkill, it's better to refer to the canonical documentation instead. We can iterate to add more detail later if requested by users.

#### Importing Specific Functions and Symbols

If you use specific functions like `data.table()`, `setDT()`, `merge()`, or special symbols like `.N`, `.SD`, `:=`, import them individually:
```r
importFrom(data.table, data.table, setDT, merge, .N, .SD, :=)
```

#### Importing S3 Generics

> If your package adds methods to any S3 generics owned by `data.table`, you need to import these generics. Check the `data.table` NAMESPACE file for the latest set of S3 generics:

Reviewer: ```suggestion
If your package adds methods to any S3 generics owned by `data.table`, you need to import these generics. Check the `data.table` NAMESPACE file for the current set of S3 generics:
```

Reviewer: NVM, see comment above

```r
importFrom(data.table, as.data.table, groupingsets, as.IDate)
```

#### Importing S4 Classes

> If your package uses S4 classes from `data.table`, you need to import these classes. This is less common but necessary for advanced use cases:

Reviewer: ```suggestion
If your package uses S4 classes from `data.table`, you need to import them. Check the `data.table` NAMESPACE file for the current set of S4 generics:
```

Reviewer: NVM, see comment above

```r
importClassesFrom(data.table, data.table, IDate, ITime)
```

#### Blanket Import

Alternatively, you can import all functions from `data.table` at once, though this is generally not recommended:

```r
import(data.table)
```

**Justification for Avoiding Blanket Imports:**
1. **Documentation**: The NAMESPACE file can serve as good documentation of how you depend on certain packages.

> 2. **Avoiding Conflicts**: Blanket imports leave you open to subtle breakage. For example, if you import(pkgA) and import(pkgB), but later pkgB exports a function also exported by pkgA, this will break your package due to conflicts in your namespace, which is disallowed by `R CMD check` and CRAN.

Reviewer: ```suggestion
2. **Avoiding Conflicts**: Blanket imports leave you open to subtle breakage. For example, if you `import(pkgA)` and `import(pkgB)`, but later pkgB exports a function also exported by pkgA, this will break your package due to conflicts in your namespace, which is disallowed by `R CMD check` and CRAN.
```

### Step 3: Update Your R code Files outside package's R/ directory

> When you move a package from `Depends` to `Imports`, it will no longer be automatically attached when your package is loaded. However, if you correctly import `data.table` in your NAMESPACE, you should not need to make changes to your package's R code files. This is different from user scripts or vignettes, which may need adjustments.

Reviewer: ```suggestion
When you move a package from `Depends` to `Imports`, it will no longer be automatically attached when your package is loaded. This can be important for examples, tests, vignettes, and demos, where `Imports` packages need to be attached explicitly.
```

#### For User Scripts and Vignettes

> If you are working with user scripts, vignettes, or other R files outside the package's R/ directory (e.g., tests and demos), you might need to explicitly load `data.table`.

Reviewer: I think these 3 lines can be removed

Author: done 👍

**Before (with `Depends`):**
```r
# data.table functions are directly available
library(MyPkgDependsDataTable)
dt <- data.table(x = 1:10, y = letters[1:10])
setDT(dt)
result <- merge(dt, other_dt, by = "x")
```

**After (with `Imports`):**
```r
# Explicitly load data.table in user scripts or vignettes
library(data.table)
library(MyPkgDependsDataTable)
dt <- data.table(x = 1:10, y = letters[1:10])
setDT(dt)
result <- merge(dt, other_dt, by = "x")
```

### Additional Considerations

- **Special Symbols**: Ensure `data.table` symbols like `.N`, `.SD`, and `:=` are imported.
- **Function Usage**: Use tools like `codetools` or `devtools` to analyze function usage and ensure all necessary imports are covered.

> - **S3 and S4 Methods**: Ensure all necessary generics and classes are imported if your package extends `data.table` functionality with custom methods.

Reviewer: I'm not sure this section adds anything to the above, WDYT about removing it?

Author: sure 👍

Some tiny parts of `data.table` C code were isolated from the R C API and can now be used from non-R applications by linking to .so / .dll files. More concrete details about this will be provided later; for now you can study the C code that was isolated from the R C API in [src/fread.c](https://github.com/Rdatatable/data.table/blob/master/src/fread.c) and [src/fwrite.c](https://github.com/Rdatatable/data.table/blob/master/src/fwrite.c).

## How to convert your Depends dependency on data.table to Imports

To convert a `Depends` dependency on `data.table` to an `Imports` dependency in an R package, follow these steps:

### 1. Update the DESCRIPTION File
In your package's `DESCRIPTION` file, change the entry from `Depends` to `Imports`. This involves moving `data.table` from the `Depends` field to the `Imports` field.

**Before:**
```r
Depends:
    R (>= 3.5.0),
    data.table
Imports:
```

**After:**
```r
Depends:
    R (>= 3.5.0)
Imports:
    data.table
```

### 2. Modify the NAMESPACE File
Ensure that any functions from `data.table` that you use in your package are imported. This can be done either by explicitly importing specific functions or by importing the entire package.

> #### Importing Specific Functions

Reviewer 2: please mention that importing specific functions should be preferred over importing the whole package

Reviewer: Covered in the above sentence; resolving

If you use specific functions like `data.table()`, `setDT()`, or `merge()`, you can import them individually in the `NAMESPACE` file:

```r
importFrom(data.table, data.table, setDT, merge)
```

#### Bulk Import

> Alternatively, you can import all functions from `data.table` at once, though this is generally not recommended as best practice:

Reviewer 2: yes this is good to say it is not recommended as best practice

```r
import(data.table)
```

By including these import statements in your `NAMESPACE` file, you ensure that the necessary functions are available for your package to use.

> ### 3. Update Your R Scripts

Reviewer 2: R scripts -> R code files

When you move a package from `Depends` to `Imports`, it will no longer be automatically attached when your package is loaded. This means you need to refer to `data.table` functions explicitly using the `data.table::` prefix, or import the functions in your scripts.

**Before (with `Depends`):**
```r
# data.table functions are directly available
dt <- data.table(x = 1:10, y = letters[1:10])
setDT(dt)
result <- merge(dt, other_dt, by = "x")
```

**After (with `Imports`):**
```r
# Use the data.table:: prefix or import functions explicitly
dt <- data.table::data.table(x = 1:10, y = letters[1:10])
data.table::setDT(dt)
result <- data.table::merge(dt, other_dt, by = "x")
```

Alternatively, you can import specific functions at the top of your R scripts:

> 

Reviewer 2: I don't think this section is correct.
You don't need the data.table::fun syntax if you import it in NAMESPACE
you would need it if you don't have an import, and are using data.table in Suggests
please revise

Author: Yes this section were meant for the packages who don't import `data.table` in their namespace but only mention them in description like this [one](https://github.com/entjos/TreeMineR/pull/2), but yeah i should change the text above to mention this 👍

Some of internally used C routines are now exported on C level thus can be used in R packages directly from their C code. See [`?cdt`](https://rdatatable.gitlab.io/data.table/reference/cdt.html) for details and [Writing R Extensions](https://cran.r-project.org/doc/manuals/r-release/R-exts.html) _Linking to native routines in other packages_ section for usage.

## Importing from non-r Applications {non-r-api}

Some tiny parts of `data.table` C code were isolated from the R C API and can now be used from non-R applications by linking to .so / .dll files. More concrete details about this will be provided later; for now you can study the C code that was isolated from the R C API in [src/fread.c](https://github.com/Rdatatable/data.table/blob/master/src/fread.c) and [src/fwrite.c](https://github.com/Rdatatable/data.table/blob/master/src/fwrite.c).

## How to convert your Depends dependency on data.table to Imports

> To convert a `Depends` dependency on `data.table` to an `Imports` dependency in an R package, follow these steps:

Reviewer: Here we use generic "an R package", below we switch to more conversational "your package", please be consistent (I prefer the latter)

> ### Step 0.  Ensure your package is passing R CMD check initially

Reviewer: ```suggestion
### Step 0. Ensure your package is passing R CMD check initially
```

> In your package's `DESCRIPTION` file, change the entry from `Depends` to `Imports`. This involves moving `data.table` from the `Depends` field to the `Imports` field.

Reviewer: IMO this section is overkill, if package author cannot grasp "Switch data.table to the Imports entry in DESCRIPTION", there are bound to be much deeper problems not solved by the later sections. I think we can assume the part this simple can be understood in one sentence, e.g.

> ### 1. Update the DESCRIPTION file to put data.table in Imports, not Depends

**Before:**
```r
Depends:
    R (>= 3.5.0),
    data.table
Imports:
```

**After:**
```r
Depends:
    R (>= 3.5.0)
Imports:
    data.table
```

### 2. Modify the NAMESPACE File

> Ensure that any functions from `data.table` that you use in your package are imported. This can be done either by explicitly importing specific functions or by importing the entire package but specific imports are considered best practice.

Reviewer: This section would frustrate me -- if I wrote a package with many files and 1000s of lines of code, how can I be sure I've imported _all_ the functions I've used? The description here sweeps the most difficult part of this step under the rug.

In the simple case of moving one package from Depends to Imports, I think the easiest solution is to just run `R CMD check` after step 1, _assuming your package is already passing `R CMD check` before starting. (So, let's make that "Step 0" -- Ensure your package is passing `R CMD check` initially).

Any symbols in your sources that are no longer defined will be identified by `R CMD check`.

Reviewer: Note that this will also help with the other thing not called out by this section yet -- the need to import data.table's "special symbols" like `.N`, `.SD`, `:=`, etc. The same `R CMD check` approach will help there, but it deserves special flagging given how common this is in data.table code.

#### Importing Specific Functions
If you use specific functions like `data.table()`, `setDT()`, or `merge()`, you can import them individually in the `NAMESPACE` file:

```r
importFrom(data.table, data.table, setDT, merge)
```

> 

Reviewer: Note that this is not complete. Users will _also_ need to make sure they `importFrom(data.table, S3Generics)` for any S3 generic owned by data.table for which they are adding a method in their package.

Readers should check our `NAMESPACE` file for the latest set of S3 generics (is there a way to get this dynamically? I think so but forget how off the top of my head), e.g. `as.data.table()`, `groupingsets()`, `as.IDate()`, etc.

Reviewer: More generally, they'd need to do other stuff for S4 methods, which they might need since we do `exportClasses()`:

https://github.com/Rdatatable/data.table/blob/6a4bd543a62cc4cee7b9bc20cb8afbcb13732091/NAMESPACE#L5

But that should be rare -- I don't see anyone doing that specifically on CRAN. I _do_ see some people importing `data.table` as a class, which I'm not quite sure I understand:

https://github.com/search?q=org%3Acran%20%2FimportClassesFrom%5B(%5Ddata%5B.%5Dtable%2F&type=code

> ### Step 2.2: Modify the NAMESPACE File

Reviewer: ```suggestion
### Step 2.2: Modify the NAMESPACE file
```

> Alternatively, you can import all functions from `data.table` at once, though this is generally not recommended as best practice:

Reviewer: It will be good to justify this a bit. Here are my top 2 reasons:

 1. The NAMESPACE can serve as good documentation of _how_ you depend on certain packages
 2. Blanket `import()` leaves you open to subtle breakage, e.g. you `import(pkgA); import(pkgB)`, but later `pkgB` exports a function also exported by `pkgA`. This will break your package because it will cause a conflict in your namespace which is disallowed by `R CMD check` (and hence CRAN).

There might be more reasons you can add. I just don't like reading "it's not best practice" claims without any hint at _why_.

Based on the `R CMD check` results, ensure all used functions, special symbols, S3 generics, and S4 classes from `data.table` are imported.

That means adding `importFrom(data.table, ...)` directives for symbols, functions, and S3 generics, and/or `importClassesFrom(data.table, ...)` directives for S4 classes as appropriate. See 'Writing R Extensions' for full details on how to do so properly.

> #### Blanket Import

Reviewer: ```suggestion
#### Blanket import
```

### 3. Update Your R code files

> When you move a package from `Depends` to `Imports`, it will no longer be automatically attached when your package is loaded. This means you need to refer to `data.table` functions explicitly using the `data.table::` prefix, or import the functions in your scripts. However, if you import `data.table` in your Namespace, you won't need to make any changes in your R scripts.

Reviewer 2: this is still incorrect.
did you copy this information from somewhere?
I don't think any changes should be necessary in R code files actually. (as long as data.table is imported)

Reviewer: Not sure what you mean here Toby. I think Nitish is talking about an analysis script, not package code. E.g.

```r
library(MyPkgDependsDataTable)
data.table(a = 1, b = MyFunction(2))
```

That script runs perfectly fine under `Depends` mode, but will stop working when switching to `Imports`.

That said, I think the easiest solution is to simply attach data.table explicitly:

```r
library(data.table)
library(MyPkgDependsDataTable)
data.table(a = 1, b = MyFunction(2))
```

This is much easier when doing one package at a time. It can be a real bear trap when switching multiple packages out of your `Depends` at once, because there can be subtle assumptions about the attach order (resulting `search()` path) that we'd need to respect to be _100%_ back-compatible. That problem is very difficult, but especially because data.table doesn't itself have any dependencies, I think we can avoid it in this discussion.

Reviewer 2: if we talk about an R analysis script (not R package code) then that needs to be clear, because the other discussion was about converting an R package

Reviewer: NB: it's still relevant in this vignette because package vignettes will be affected. So will other R files outside R/ directory like tests and demos

```r
import(data.table)
```

**Justification for Avoiding Blanket Imports:**
1. **Documentation**: The NAMESPACE file can serve as good documentation of how you depend on certain packages.
2. **Avoiding Conflicts**: Blanket imports leave you open to subtle breakage. For example, if you `import(pkgA)` and `import(pkgB)`, but later pkgB exports a function also exported by pkgA, this will break your package due to conflicts in your namespace, which is disallowed by `R CMD check` and CRAN.

> ### Step 3: Update Your R code Files outside package's R/ directory

Reviewer: ```suggestion
### Step 3: Update Your R code files outside the package's R/ directory
```

```r
# Use the data.table:: prefix or import functions explicitly
dt <- data.table::data.table(x = 1:10, y = letters[1:10])
data.table::setDT(dt)
result <- data.table::merge(dt, other_dt, by = "x")
```

Alternatively, you can import specific functions at the top of your R scripts:

```r
# Import specific functions at the beginning of the script
data.table::data.table
data.table::setDT
data.table::merge

# Now you can use them without the data.table:: prefix
dt <- data.table(x = 1:10, y = letters[1:10])
setDT(dt)
result <- merge(dt, other_dt, by = "x")
```

**Important**: To avoid the need for changes in your R scripts, ensure you import `data.table` in your Namespace. This allows the functions to be available as they were with `Depends`.

### Benefits of Using `Imports`

> - **Namespace Management**: Only the functions your package explicitly imports are available, reducing the risk of function name clashes.

Reviewer: It is also much easier for maintenance, as mentioned also above as your upstream dependencies' own APIs evolve, relying too much on `Depends` is bound to cause headaches & conflicts.

> ### Benefits of Using `Imports`

Reviewer: ```suggestion
### Benefits of using `Imports`
```

- **Easier Maintenance**: It simplifies maintenance tasks as upstream dependencies' APIs evolve. Depending too much on `Depends` can lead to conflicts and compatibility issues over time.

> 
> By following these steps, you can convert your package's dependency on `data.table` from `Depends` to `Imports`, leading to better practice in R package development.

Reviewer: ```suggestion
```

> ## Importing from non-R Applications {non-r-API}

Reviewer 3: I see some words being capitalized in the headings apart from the first word or proper nouns ('Applications' here, and more below), avoid that

## How to convert your Depends dependency on data.table to Imports

To convert a `Depends` dependency on `data.table` to an `Imports` dependency in an R package, follow these steps:

### 1. Update the DESCRIPTION File
In your package's `DESCRIPTION` file, change the entry from `Depends` to `Imports`. This involves moving `data.table` from the `Depends` field to the `Imports` field.

**Before:**
```r
Depends:
    R (>= 3.5.0),
    data.table
Imports:
```

**After:**
```r
Depends:
    R (>= 3.5.0)
Imports:
    data.table
```

### 2. Modify the NAMESPACE File
Ensure that any functions from `data.table` that you use in your package are imported. This can be done either by explicitly importing specific functions or by importing the entire package.

#### Importing Specific Functions
If you use specific functions like `data.table()`, `setDT()`, or `merge()`, you can import them individually in the `NAMESPACE` file:

```r
importFrom(data.table, data.table, setDT, merge)
```

#### Bulk Import

> Alternatively, you can import all functions from data.table at once, though this is generally not recommended for best practices:

Reviewer 3: `data.table`
as best practice*

Author: done 👍
